pkg/apiserver/experiment/experiment.go

```diff
@@ -27,1290 +26,579 @@ import (
-		}
-
-		return Detail{}, err
-	}
-
-	gvk, err := apiutil.GVKForObject(chaos, s.scheme)
+// @Summary Create a new chaos experiment.
+// @Description Pass a JSON object to create a new chaos experiment. The schema for JSON is the same as the YAML schema for the Kubernetes object.
+// @Tags experiments
+// @Accept json
+// @Produce json
+// @Param chaos body map[string]interface{} true "the chaos definition"
+// @Success 200 {object} map[string]interface{}
+// @Failure 400 {object} utils.APIError
+// @Failure 500 {object} utils.APIError
+// @Router /experiments [post]
+func (s *Service) create(c *gin.Context) {
+	kubeCli, err := clientpool.ExtractTokenAndGetClient(c.Request.Header)
 	if err != nil {
-		return Detail{}, err
-	}
-
-	return Detail{
-		Experiment: Experiment{
-			Base: Base{
-				Kind:      gvk.Kind,
-				Namespace: chaos.Namespace,
-				Name:      chaos.Name,
-			},
-			UID:     string(chaos.GetUID()),
-			Created: chaos.GetCreationTimestamp().Format(time.RFC3339),
-			Status:  string(utils.GetChaosState(chaos)),
-		},
-		KubeObject: core.KubeObjectDesc{
-			TypeMeta: metav1.TypeMeta{
-				APIVersion: gvk.GroupVersion().String(),
-				Kind:       gvk.Kind,
-			},
-			Meta: core.KubeObjectMeta{
-				Name:        chaos.Name,
-				Namespace:   chaos.Namespace,
-				Labels:      chaos.Labels,
-				Annotations: chaos.Annotations,
-			},
-			Spec: chaos.Spec,
-		},
-	}, nil
-}
-
-func (s *Service) getNetworkChaosDetail(namespace string, name string, kubeCli client.Client) (Detail, error) {
-	chaos := &v1alpha1.NetworkChaos{}
-
-	chaosKey := types.NamespacedName{Namespace: namespace, Name: name}
-	if err := kubeCli.Get(context.Background(), chaosKey, chaos); err != nil {
-		if apierrors.IsNotFound(err) {
-			return Detail{}, utils.ErrNotFound.NewWithNoMessage()
-		}
+		u.SetAPIError(c, u.ErrBadRequest.WrapWithNoMessage(err))
 
-		return Detail{}, err
+		return
 	}
 
-	gvk, err := apiutil.GVKForObject(chaos, s.scheme)
-	if err != nil {
-		return Detail{}, err
+	var exp map[string]interface{}
+	if err = u.ShouldBindBodyWithJSON(c, &exp); err != nil {
+		return
 	}
+	kind := exp["kind"].(string)
 
-	return Detail{
-		Experiment: Experiment{
-			Base: Base{
-				Kind:      gvk.Kind,
-				Namespace: chaos.Namespace,
-				Name:      chaos.Name,
-			},
-			UID:     string(chaos.GetUID()),
-			Created: chaos.GetCreationTimestamp().Format(time.RFC3339),
-			Status:  string(utils.GetChaosState(chaos)),
-		},
-		KubeObject: core.KubeObjectDesc{
-			TypeMeta: metav1.TypeMeta{
-				APIVersion: gvk.GroupVersion().String(),
-				Kind:       gvk.Kind,
-			},
-			Meta: core.KubeObjectMeta{
-				Name:        chaos.Name,
-				Namespace:   chaos.Namespace,
-				Labels:      chaos.Labels,
-				Annotations: chaos.Annotations,
-			},
-			Spec: chaos.Spec,
-		},
-	}, nil
-}
+	if chaosKind, ok := v1alpha1.AllKinds()[kind]; ok {
+		reflect.ValueOf(chaosKind.Chaos).Elem().FieldByName("ObjectMeta").Set(reflect.ValueOf(metav1.ObjectMeta{}))
```

Comment:
Because the previous chaos' `ObjectMeta` is still in `chaosKind.Chaos`, this needs to be cleared to prevent errors from occurring (The key thing is to remove the `ResourceVersion`).

Reply:
using DeepCopy could prevent this “previous object“ issue

Reply:
Actually, each usage of `v1alpha1.AllKind()` should be wrapped with `DeepCopyObject`. I think we should provide the easy-to-use and more obviously method name.

ref: https://github.com/chaos-mesh/chaos-mesh/pull/2193#discussion_r690915555

Reply:
or call `DeepCopyObject` or `DeepCopyList` inside of `clone()`, but it is not so "obvious".